### ISSUE TITLE
Key-Value Logging instead of LogEvent

### ISSUE THREAD

**Issue by cwe1ss** [opentracing/opentracing-csharp#9]
General description and motivation: https://github.com/opentracing/opentracing.github.io/issues/96
(Ongoing) discussion in the java repository: https://github.com/opentracing/opentracing-java/pull/47

I'd propose the following signature changes in C# for key value logging:

**Remove**: (noone is using the library yet)
```csharp
// ISpan interface
ISpan LogEvent(string eventName, object payload = null);
ISpan LogEvent(DateTimeOffset timestamp, string eventName, object payload = null);
```

**Add**:
```csharp

// custom type because using KeyValuePair<string, object> or Tuple<string, object> is annoying.
public struct LogValue
{
  public string Key { get; }
  public object Value { get; }
  public LogValue(string key, object value) { ... } // constructor
}

// ISpan interface
ISpan Log(params LogValue[] fields);
ISpan Log(DateTimeOffset timestamp, params LogValue[] fields);
```

**Background:**
I created a small performance test and compared `params object[]`, `params LogValue[]`, `IList<KeyValuePair<string, object>>`, `IDictionary<string, object>`. Results may vary, so this should just be an indication about the difference between the different methods. More details about this can be found in the mentioned Java issue.

```
Iterations: 10,000,000
Log(object[]): 633ms
Log(LogValue[]): 839ms
Log(KVP List): 1456ms
Log(Dictionary): 3956ms
```
The Java implementation uses an object-array right now, however I don't like using this in C# for the following reasons:
* Users can pass non-string keys
* Users can pass an uneven number of arguments (they might miss a key/value)
* Tracers can't easily loop through all fields (they have to check every second entry to get the keys)

Using a custom type solves these issues and has only marginally worse performance. It also keeps the door open to use inheritance in the future (e.g. for lazy evaluated values, ...)

What we don't have with using an array instead of a dictionary are unique keys, but given the performance impact I think this is an acceptable compromise.

**Comment by dawallin** [comment]
@cwe1ss Have you looked at [BenchmarkDotNet ](https://github.com/PerfDotNet/BenchmarkDotNet). Perfect for doing these kind of benchmarks. 

I tried and get the same trends as you show here (custom type are acctually faster than object array if you run release mode instead of debug as you edited in the Java issue). Only one comment, in the Log(KVP List) it is the list that is the slow part. Using KeyValuePair<string, object> is almost as fast as the LogValue custom type. 

I think what you suggest is a good way to go, much better than a generic object array, and much faster than a dictionary.

**Comment by cwe1ss** [comment]
The Java PR opentracing/opentracing-java#47 now also uses a custom type. I will create a C# PR with an equivalent solution.

One thing I'm not sure about is whether C# can also do something like `Span log(LogField<?>... fields);`. I think in C#, generics always have to be defined, meaning a list can not contain elements with different generic types. 
Most probably this means, we'll have to use `object` as the type for the value.

**Comment by cwe1ss** [comment]
closed because the actual PR #13 contains the latest information.
